pkg/cmd/pulumi/config.go

```diff
@@ -80,10 +80,188 @@ func newConfigCmd() *cobra.Command {
+	configNamespace string
+	value           string
+	encrypted       bool
+}
+
+func newConfigCopyCmd(stack *string) *cobra.Command {
+	var path bool
+	var newStackName string
+
+	cpCommand := &cobra.Command{
+		Use:   "cp <key>",
+		Short: "Copy config to another stack",
+		Long: "Copies the config from the current stack to the newly specified stack. If a key is omitted,\n" +
+			"then all of the config of the current stack will be copied to the new stack.",
+		Args: cmdutil.MaximumNArgs(1),
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			opts := display.Options{
+				Color: cmdutil.GetGlobalColorization(),
+			}
+
+			s, err := requireStack(*stack, false, opts, true /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+
+			if s.Ref().Name().String() == newStackName {
+				return errors.New("source and destination stacks must be different when copying config")
+			}
+
+			currentStack, err := loadProjectStack(s)
+			if err != nil {
+				return err
+			}
+
+			configToCopy, err := buildConfigKeysToCopy(args, path, s, currentStack)
+			if err != nil {
+				return err
+			}
+
+			ns, err := requireStack(newStackName, false, opts, false /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+			newStack, err := loadProjectStack(ns)
+			if err != nil {
+				return err
+			}
+
+			var projectNeedsSaved bool
+			for _, key := range configToCopy {
+				err := buildAndSaveConfig(ns, newStack, key)
+				if err != nil {
+					return err
+				}
+				projectNeedsSaved = true
+			}
+
+			if projectNeedsSaved {
+				return saveProjectStack(ns, newStack)
+			}
+
+			return nil
+		}),
+	}
+
+	cpCommand.PersistentFlags().BoolVar(
+		&path, "path", false,
+		"The key contains a path to a property in a map or list to set")
+	cpCommand.PersistentFlags().StringVarP(
+		&newStackName, "new-stack", "", "",
+		"The name of the new stack to copy the config to")
+
+	return cpCommand
+}
+
+func buildConfigKeysToCopy(args []string, path bool, stack backend.Stack,
+	projectStack *workspace.ProjectStack) ([]stackConfigKeyValues, error) {
+	var decrypter config.Decrypter
+	var keysToCopy []stackConfigKeyValues
+	cfg := projectStack.Config
+	if len(args) > 0 {
+		key, err := parseConfigKey(args[0])
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid configuration key")
+		}
+
+		v, ok, err := cfg.Get(key, path)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			configToCopy := stackConfigKeyValues{
+				name:            key.Name(),
+				configNamespace: key.Namespace(),
+				encrypted:       v.Secure(),
+			}
+			if v.Secure() {
+				var err error
+				if decrypter, err = getStackDecrypter(stack); err != nil {
+					return nil, errors.Wrap(err, "could not create a decrypter")
+				}
+			} else {
+				decrypter = config.NewPanicCrypter()
+			}
+			raw, err := v.Value(decrypter)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not decrypt configuration value")
+			}
+			configToCopy.value = raw
```

Comment:
Does this preserve values that are objects (i.e. when `v.Object()` is true), or will such values end up as JSON strings in the destination stack config? I'd have to checkout these changes to see, but ideally we would preserve these so the destination YAML looks as expected.

Reply:
Just tried this and it doesn't work the way I'd expect.

```
$ pulumi stack init dev
$ pulumi stack init dev2
$ pulumi stack select dev
$ pulumi config set aws:region us-west-2
$ pulumi config set --path data.active true
$ pulumi config set --path data.nums[0] 1
$ pulumi config set --path data.nums[1] 2
$ pulumi config set --path data.nums[2] 3
$ cat Pulumi.dev.yaml
config:
  aws:region: us-west-2
  cpcfgtest:data:
    active: true
    nums:
    - 1
    - 2
    - 3
$ pulumi config cp --new-stack dev2
$ cat Pulumi.dev2.yaml
config:
  aws:region: us-west-2
  cpcfgtest:data: '{"active":true,"nums":[1,2,3]}'
```

I would expect `Pulumi.dev2.yaml` to look like `Pulumi.dev.yaml`:

```
config:
  aws:region: us-west-2
  cpcfgtest:data:
    active: true
    nums:
    - 1
    - 2
    - 3
```